### PR TITLE
[BEAM-11408] Integrate Python BigQuery sink with GroupIntoBatches

### PR DESCRIPTION
**Pull request by nehsyc**
This is a parity of Java changes (#13496 and #13859).

Use `GroupIntoBatches.WithShardedKey` API to group and batch write before streaming to BigQuery service. Currently batching is done best-effort upon bundle finalization.

This PR adds an option to `WriteToBigQuery` to toggle between the existing and new implementation for both streaming inserts and file loads.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by nehsyc** [comment]
R: @pabloem
CC: @reuvenlax

**Comment by pabloem** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 1]
```diff
@@ -1490,8 +1536,8 @@ def __init__(
       schema (str,dict,ValueProvider,callable): The schema to be used if the
         BigQuery table to write has to be created. This can be either specified
         as a :class:`~apache_beam.io.gcp.internal.clients.bigquery.\
-bigquery_v2_messages.TableSchema`. or a `ValueProvider` that has a JSON string,
-        or a python dictionary, or the string or dictionary itself,
+        bigquery_v2_messages.TableSchema`. or a `ValueProvider` that has a JSON
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/bigquery.py:1539]: I think this may be causing the docs test failure

**Review by pabloem — COMMENTED** [review]
There's some lint issues here: https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Commit/8336/console

Specifically

```
00:02:47.755 apache_beam/io/gcp/bigquery.py:1363:0: C0301: Line too long (102/80) (line-too-long)
00:02:47.755 apache_beam/io/gcp/bigquery.py:1361:0: W0105: String statement has no effect (pointless-string-statement)
```

I think you can just document those constants with the usual `# comment` instead of the `"""" comment """` ones.

Also, I guess you've figured this out, but it will be good if you add an integration test. You can find those mainly in bigquery_test.py, they're tagged with `@attr('IT')`

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 1 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/gcp/bigquery.py:1539]: From the test suite:
```
00:02:30.217 /home/jenkins/jenkins-slave/workspace/beam_PreCommit_PythonDocs_Commit/src/sdks/python/test-suites/tox/pycommon/build/srcs/sdks/python/apache_beam/io/gcp/bigquery.py:docstring of apache_beam.io.gcp.bigquery.WriteToBigQuery:28: WARNING: py:class reference target not found: apache_beam.io.gcp.internal.clients.bigquery.        bigquery_v2_messages.TableSchema
```

Comment (nehsyc) [reply sdks/python/apache_beam/io/gcp/bigquery.py]: Thanks for pointing it out! Fixed and pushed a new commit.

**Comment by pabloem** [comment]
Run Python 3.8 PostCommit

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 2]
```diff
@@ -1759,6 +1814,8 @@ def serialize(side_inputs):
         'triggering_frequency': self.triggering_frequency,
         'validate': self._validate,
         'temp_file_format': self._temp_file_format,
+        'ignore_insert_ids': self._ignore_insert_ids,
```
Comment (nehsyc) [inline sdks/python/apache_beam/io/gcp/bigquery.py:1817]: `ignore_insert_ids` was previously not added to runner api parameter. Wondering if this is a bug as when we convert from the runner api the parameter will be lost.

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py [hunk 3]
```diff
@@ -716,8 +826,10 @@ def test_multiple_destinations_transform(self):
               max_file_size=20,
               max_files_per_bundle=-1))
 
+  @parameterized.expand(
```
Comment (nehsyc) [inline sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py:829]: I was not able to run this with Dataflow runner. I can revert it back if more appropriate.

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 2 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/gcp/bigquery.py:1817]: ah thanks Siyuan!

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 4]
```diff
@@ -1467,7 +1511,8 @@ def __init__(
       triggering_frequency=None,
       validate=True,
       temp_file_format=None,
-      ignore_insert_ids=False):
+      ignore_insert_ids=False,
+      with_auto_sharding=False):
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/bigquery.py:1515]: let's have a JIRA issue and a `TODO(BEAM-XXXX)` to track making this the default? : )

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 5]
```diff
@@ -1403,27 +1417,57 @@ def expand(self, input):
         retry_strategy=self.retry_strategy,
         test_client=self.test_client,
         additional_bq_parameters=self.additional_bq_parameters,
-        ignore_insert_ids=self.ignore_insert_ids)
+        ignore_insert_ids=self.ignore_insert_ids,
+        with_batched_input=(
+            not self.ignore_insert_ids and self.with_auto_sharding))
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/bigquery.py:1422]: how come `ignore_insert_ids` affects this parameter?

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 6]
```diff
@@ -1254,13 +1257,20 @@ def process(self, element, *schema_side_inputs):
...
+      self._total_buffered_rows += 1
+      if len(self._rows_buffer[destination]) >= self._max_batch_size:
+        return self._flush_batch(destination)
+      elif self._total_buffered_rows >= self._max_buffered_rows:
+        return self._flush_all_batches()
+    else:
+      # The input is already batched per destination, flush the rows now.
+      batched_rows = element[1]
+      for row in batched_rows:
+        self._rows_buffer[destination].append(row)
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/bigquery.py:1272]:
```suggestion
      self._rows_buffer[destination].extend(batched_rows)
```

Comment (nehsyc) [reply sdks/python/apache_beam/io/gcp/bigquery.py]: Neat!

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 5 above)
Comment (nehsyc) [reply sdks/python/apache_beam/io/gcp/bigquery.py]: Good catch!! Removed.

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 4 above)
Comment (nehsyc) [reply sdks/python/apache_beam/io/gcp/bigquery.py:1514]: Done.

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py (hunk 3 above)
Comment (nehsyc) [reply sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py]: I reverted the test as I still couldn't run it successfully even with local Dataflow runner. I will follow up in a separate PR if needed.

**Comment by nehsyc** [comment]
Thanks for the review!

I got the following lint errors which are supposed to be suppressed. Any ideas?
```
17:47:48 ERROR: /home/jenkins/jenkins-slave/workspace/beam_PreCommit_PythonLint_Commit/src/sdks/python/test-suites/tox/py37/build/srcs/sdks/python/apache_beam/io/gcp/bigquery_test.py Imports are incorrectly sorted.
17:47:48 --- /home/jenkins/jenkins-slave/workspace/beam_PreCommit_PythonLint_Commit/src/sdks/python/test-suites/tox/py37/build/srcs/sdks/python/apache_beam/io/gcp/bigquery_test.py:before	2021-02-23 01:43:30.069624
17:47:48 +++ /home/jenkins/jenkins-slave/workspace/beam_PreCommit_PythonLint_Commit/src/sdks/python/test-suites/tox/py37/build/srcs/sdks/python/apache_beam/io/gcp/bigquery_test.py:after	2021-02-23 01:47:43.161598
17:47:48 @@ -38,6 +38,10 @@
17:47:48  import mock
17:47:48  import pytz
17:47:48  from nose.plugins.attrib import attr
17:47:48 +# Protect against environments where bigquery library is not available.
17:47:48 +# pylint: disable=wrong-import-order, wrong-import-position
17:47:48 +from parameterized import param
17:47:48 +from parameterized import parameterized
```

**Comment by pabloem** [comment]
I think you can just move lines 80-81 to be 41-42, and that should be enough?

**Comment by pabloem** [comment]
Run Python 3.8 PostCommit

**Comment by pabloem** [comment]
Run Python 3.8 PostCommit

**Comment by pabloem** [comment]
thanks @nehsyc ! this will be in Beam 2.29.0 - excited to set it to the wild : )

**Pull request merged**